### DISCUSSION TITLE
Params T

### DISCUSSION THREAD

**Discussion by TonyValenti** [dotnet/csharplang#3573] [Language Ideas]
Please allow any type that implements ````Add```` (either directly or as an extension method) and has a default constructor to be used as a Params type.

for example:
````
public void Test1(params List<string> Values) { }
public void Test2(params HashSet<string> Values) { }
public void Test3(params LinkedList<string> Values) { }

public static void Add<T>(this LinkedList<T> This, T Item){
  This.AddList(Item);
}

...

Test1("Hello", "World");
/*
This becomes:
var x = new List<string>();
x.Add("Hello");
x.Add("World");
Test1(x);

etc.
*/


````

**Comment by jnm2** [comment]
But also take into account methods like ImmutableArray<T>.Add which return an updated copy. You want to use a builder in that case.

**Comment by svick** [comment]
Closely related proposal: https://github.com/dotnet/csharplang/issues/179 and https://github.com/dotnet/csharplang/issues/178.

**Comment by miloush** [comment]
While related, this is a very different proposal. The original ones are about accepting additional, already instantiated types, while this one is about constructing new instances to be passed to the method.

I am not very keen on this one as you can't really see from the code that something is being instantiated and what.

Moreover, what is this proposal thinking about ambiguities?

```C#
void Test(params List<string> values) {}
void Test(params HashSet<string> values) {}
void Test(params LinkedList<string> values) {}

Test("Hello", "World"); // what happens now?
```

**Comment by CyrusNajmabadi** [comment]
> what happens now?

You get an ambiguity error.

**Comment by miloush** [comment]
@CyrusNajmabadi Ambiguity error can usually be resolved, e.g. by explicit casting, this one does not seem to be resolvable.

**Comment by miloush** [comment]
Correction, unless you are willing to give up the params syntax, then you could do `Test(new List<string> { "Hello", "World" })`...
